sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala

```diff
@@ -606,9 +606,17 @@ abstract class HadoopFsRelation private[sql](
         // we need to cast into the data type that user specified.
         def castPartitionValuesToUserSchema(row: InternalRow) = {
           InternalRow((0 until row.numFields).map { i =>
-            Cast(
-              Literal.create(row.getString(i), StringType),
-              userProvidedSchema.fields(i).dataType).eval()
+            row.isNullAt(i) match {
```

Comment:
The null-checking is cheap.

Or we could use `Literal.create(row.getUTF8String(i), StringType)`, to avoid the conversion between String and UTF8String.